src/aggregator/client/queue.go

```diff
@@ -204,22 +223,61 @@ func (q *queue) Close() error {
 	return nil
 }
 
-func (q *queue) drain() {
+func (q *queue) drain(tC <-chan (time.Time)) {
 	defer q.wg.Done()
+	defer q.conn.Close()
+	lastDrain := time.Now()
+
+	for {
+		select {
+		case qitem := <-q.bufCh:
+			var needFlush bool
+			msg := qitem.Bytes()
+			q.buf = append(q.buf, msg...)
+			qitem.Close()
+
+			if len(q.buf) > q.batchSize {
+				needFlush = true
+			} else {
+				needFlush = time.Since(lastDrain) >= q.writeInterval
+			}
 
-	for buf := range q.bufCh {
-		if err := q.writeFn(buf.Bytes()); err != nil {
-			q.log.Error("write data error",
-				zap.String("instance", q.instance.Endpoint()),
-				zap.Error(err),
-			)
-			q.metrics.connWriteErrors.Inc(1)
-		} else {
-			q.metrics.connWriteSuccesses.Inc(1)
+			if !needFlush {
+				continue
+			}
+
+			if err := q.writeFn(q.buf); err != nil {
+				q.log.Error("error writing data", zap.String("instance", q.instance.Endpoint()), zap.Error(err))
+				q.metrics.connWriteErrors.Inc(1)
+			} else {
+				q.metrics.connWriteSuccesses.Inc(1)
+			}
+			qitem.Close()
+
+			q.buf = q.buf[:0]
+			lastDrain = time.Now()
+		case ts := <-tC:
+			delta := ts.Sub(lastDrain)
+			lastDrain = ts
+			if delta < q.writeInterval || len(q.buf) == 0 {
+				continue
```

Comment:
So only problem I see here using a ticker, is that if you wake up and you recently flushed because of size limit so you `continue` you can potentially sleep a whole other period (i.e. 100ms) before you potentially send again (i.e. you send the next batch 150ms later if you sent a packet 50ms into the first 100ms period and you never wrote anything afterwards).

I know you cover some of this case with `needFlush = time.Since(lastDrain) >= q.writeInterval` when you receive a buffer, however if you never receive another buffer and you didn't hit the size limit before you begin the second minWriteInterval time period then it doesn't get sent until next ticker tick.

So worst case is you approach 2x minWriteInterval for this pathological case.  It's probably not that bad, I just wanted to call it out.

The other option is to subtract the difference when when it was last drained, then specifically sleep for minWriteInterval-delta and wake up again.  That way you are putting a max deadline on when data is sent.

Either way, just wanted to document the differences between the two strategies.